test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/infinispan/infinispan-operator/controllers"
+	"github.com/infinispan/infinispan-operator/controllers/constants"
+	batchtest "github.com/infinispan/infinispan-operator/test/e2e/batch"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	coreos "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	testKube = tutils.NewTestKubernetes(os.Getenv("TESTING_CONTEXT"))
+	// startChannel          = os.Getenv("SUBSCRIPTION_CHANNEL_START")
+	subName               = constants.GetEnvWithDefault("SUBSCRIPTION_NAME", "infinispan-operator")
+	subNamespace          = constants.GetEnvWithDefault("SUBSCRIPTION_NAMESPACE", tutils.Namespace)
+	sourceChannel         = constants.GetEnvWithDefault("SUBSCRIPTION_CHANNEL_SOURCE", "2.1.x")
+	latestChannel         = constants.GetEnvWithDefault("SUBSCRIPTION_CHANNEL_LATEST", "2.2.x")
+	catalogSource         = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE", "test-catalog")
+	catalogSourcNamespace = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE_NAMESPACE", tutils.Namespace)
+	subPackage            = constants.GetEnvWithDefault("SUBSCRIPTION_PACKAGE", "infinispan")
+	subStartingCsv        = constants.GetEnvWithDefault("SUBSCRIPTION_STARTING_CSV", "infinispan-operator.v2.1.5")
+
+	latestImage = constants.GetEnvWithDefault("LATEST_OPERAND_IMAGE", constants.DefaultOperandImageOpenJDK)
+)
+
+func TestUpgrade(t *testing.T) {
+	testKube.NewNamespace(tutils.Namespace)
+
+	sub := &coreos.Subscription{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: coreos.SubscriptionCRDAPIVersion,
+			Kind:       coreos.SubscriptionKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      subName,
+			Namespace: subNamespace,
+		},
+		Spec: &coreos.SubscriptionSpec{
+			Channel:                sourceChannel,
+			CatalogSource:          catalogSource,
+			CatalogSourceNamespace: catalogSourcNamespace,
+			InstallPlanApproval:    coreos.ApprovalManual,
+			Package:                subPackage,
+			StartingCSV:            subStartingCsv,
+		},
+	}
+	testKube.CreateSubscription(sub)
+	defer testKube.DeleteNamespace(tutils.Namespace)
+
+	testKube.WaitForSubscriptionState(subName, subNamespace, coreos.SubscriptionStateUpgradePending)
+
+	// Upgrade the Subscription channel if required
+	if sourceChannel != latestChannel {
+		testKube.UpdateSubscriptionChannel(subName, subNamespace, latestChannel)
+	}
+
+	testKube.WaitForSubscriptionState(subName, subNamespace, coreos.SubscriptionStateUpgradePending)
+	// Approve the initial startingCSV InstallPlan
+	testKube.ApproveInstallPlan(subName, subNamespace, 1)
+
+	testKube.WaitForCrd(&apiextv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "infinispans.infinispan.org",
+		},
+	})
+
+	spec := tutils.DefaultSpec(testKube)
```

Comment:
Better would be to use at least 2 replicas sized cluster. Lot of issues could be hidden without using actual cluster